### PR TITLE
[dynamo] fix graph break, improve hygeine - enforce using ConstantVariable for `torch.device`,`torch.dtype`

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/pull/112332/files#r1375690808

Simplify code paths, fix graph break

```
torch._dynamo.exc.InternalTorchDynamoError: TorchVariable() has no type
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by jon-chuang** [comment]
Looks like this PR unblocks some compilation due to not graph breaking. We need to add `aten.cauchy.default` decomp to unblock the test

File: test/test_tensor_creation_ops.py [hunk 1]
```diff
@@ -3140,6 +3140,7 @@ class TestRandomTensorCreation(TestCase):
 
     # TODO: add torch.complex64, torch.complex128
     @dtypes(torch.float, torch.double)
+    @skipIfTorchDynamo("normal doesn't compile with non-default dtype")
```
Comment (jon-chuang) [inline test/test_tensor_creation_ops.py:3143]: fixed by https://github.com/pytorch/pytorch/pull/112467

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -1662,7 +1662,16 @@ def check_skip_condition(node, parent, is_output):
...
-    if get_decompositions([op]) and warn and bool(os.getenv("CI")):
+    if (
+        warn
+        and bool(os.getenv("CI"))
+        and get_decompositions([op])
+        # if fallback_random, we allow not decomposing random
+        and not (
+            config.fallback_random
+            and op in torch._decomp.decompositions_for_rng.extra_random_decomps
+        )
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply torch/_inductor/lowering.py:1673]: it is tangentially related - as new tests that don't graph break now can compile and will fail

inductor will complain if you fallback - but we fallback random in the tests

File: torch/_dynamo/variables/tensor.py [hunk 3]
```diff
@@ -233,9 +233,9 @@ def var_getattr(self, tx, name):
         if name == "ndim" and self.ndim is not None:
             result = ConstantVariable.create(self.ndim, **options)
         elif name == "dtype" and self.dtype is not None:
-            result = TorchVariable(self.dtype, **options)
+            result = ConstantVariable.create(self.dtype, **options)
         elif name == "device" and self.device is not None:
-            result = TorchVariable(self.device, **options)
+            result = ConstantVariable.create(self.device, **options)
         elif name == "layout" and self.layout is not None:
             result = TorchVariable(self.layout, **options)
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply torch/_dynamo/variables/tensor.py:240]: done

File: test/dynamo/test_misc.py [hunk 4]
```diff
@@ -7500,17 +7500,37 @@ def fn(target):
...
+            c = torch.zeros(2, 3, dtype=target_dtype)
+            return a + b + c
+
+        from torch._dynamo.variables import ConstantVariable
+
+        dtype = torch.float16
+        expected_variable = ConstantVariable(dtype)
+        self.assertEqual(expected_variable.python_type(), type(dtype))
+
+        opt_func = torch._dynamo.optimize("inductor")(fn)
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply test/dynamo/test_misc.py:7529]: Let me get rid of assertIsInstance

[the remark this answers is not in the document]
Comment (lezcano) [reply test/dynamo/test_misc.py:7529]: at any rate, I think looking at the number of frames + not using inductor if we are testing dynamo is a good thing regardless

[the remark this answers is not in the document]
Comment (jon-chuang) [reply test/dynamo/test_misc.py]: I use native assert now, and fullpython=True

File: torch/_dynamo/variables/torch.py [hunk 5]
```diff
@@ -282,6 +282,9 @@ def __init__(self, value, **kwargs):
         check_allowed_op(value)
         self.value = value
 
+        assert not isinstance(
+            value, (torch.dtype, torch.device)
```
Comment (lezcano) [inline torch/_dynamo/variables/torch.py:286]: + layout

File: torch/_dynamo/variables/tensor.py [hunk 6]
```diff
@@ -233,11 +233,11 @@ def var_getattr(self, tx, name):
...
             result = ConstantVariable.create(self.ndim, **options)
         elif name == "dtype" and self.dtype is not None:
-            result = TorchVariable(self.dtype, **options)
+            result = ConstantVariable.create(self.dtype, **options)
         elif name == "device" and self.device is not None:
-            result = TorchVariable(self.device, **options)
+            result = ConstantVariable.create(self.device, **options)
         elif name == "layout" and self.layout is not None:
-            result = TorchVariable(self.layout, **options)
+            result = ConstantVariable(self.layout, **options)
```
Comment (lezcano) [inline torch/_dynamo/variables/tensor.py:240]: .create

File: test/dynamo/test_misc.py [hunk 7]
```diff
@@ -7492,22 +7492,53 @@ def ff2(x):
         self.assertEqual(counter.op_count, 6)
 
     def test_torch_device_python_type(self):
+        for device, device_type, index in [
+            ("cpu", "cpu", None),
+            ("cuda:0", "cuda", 0),
+        ]:
```
Comment (lezcano) [inline test/dynamo/test_misc.py:7498]: Here we are testing inductor, so should go in the inductor folder.

**Review by lezcano — APPROVED** [review]
Three minor points, otherwise LGTM

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (jon-chuang) [reply test/dynamo/test_misc.py:7498]: Hmm, are we testing inductor though? Not sure why the original test used inductor. Personally I think it's fine to use eager.

Comment (lezcano) [reply test/dynamo/test_misc.py:7498]: then let's run it on eager

File: torch/_dynamo/variables/torch.py [hunk 8]
```diff
@@ -283,7 +283,7 @@ def __init__(self, value, **kwargs):
         self.value = value
 
         assert not isinstance(
-            value, (torch.dtype, torch.device, torch.layout)
+            value, (torch.dtype, torch.device)
```
Comment (lezcano) [inline torch/_dynamo/variables/torch.py:286]: Now layout can be both a torch variable or a ConstantVariable?? Either revert all the layout changes or fix all the issues, but don't leave divergent paths.

**Review by lezcano — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch.py (hunk 8 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/torch.py:286]: Yeah let me get rid of making it a constant variable. There are some other paths that want it to be TorchVariable

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**